File: bp-workout/Services/UserProgramLibrary.swift
import Combine
import Foundation

/// Programs the user has added to their profile (Workout tab picker). When nothing is stored yet, every catalog program is treated as included.
@MainActor
final class UserProgramLibrary: ObservableObject {
    static let shared = UserProgramLibrary()

    @Published private(set) var updateCounter: Int = 0

    private let defaultsKey = "profile.libraryProgramIds"
    private let activeProgramDefaultsKey = "workoutHub.activeProgramId"

    private init() {}

    func idsInLibrary(catalogIds: [String]) -> Set<String> {
        let catalog = Set(catalogIds)
        if UserDefaults.standard.object(forKey: defaultsKey) == nil {
            return catalog
        }
        let stored = Set(UserDefaults.standard.stringArray(forKey: defaultsKey) ?? [])
        return stored.intersection(catalog)
    }

    func isInLibrary(programId: String, catalog: [WorkoutProgram]) -> Bool {
        let ids = catalog.map(\.id)
        return idsInLibrary(catalogIds: ids).contains(programId)
    }

    func setProgramInLibrary(_ programId: String, enabled: Bool, catalog: [WorkoutProgram]) {
        let catalogIds = catalog.map(\.id)
        var next = idsInLibrary(catalogIds: catalogIds)
        if enabled {
            next.insert(programId)
        } else {
            next.remove(programId)
        }
        UserDefaults.standard.set(Array(next), forKey: defaultsKey)
        updateCounter += 1
        reconcileStoredActiveProgram(catalog: catalog)
    }

    /// Keeps `workoutHub.activeProgramId` pointing at a program still in the library (or clears it).
    private func reconcileStoredActiveProgram(catalog: [WorkoutProgram]) {
        let inLib = programsInProfile(from: catalog)
        let d = UserDefaults.standard
        if inLib.isEmpty {
            d.set("", forKey: activeProgramDefaultsKey)
            return
        }
        if let saved = d.string(forKey: activeProgramDefaultsKey), !saved.isEmpty,
           inLib.contains(where: { $0.id == saved }) {
            return
        }
        d.set(inLib[0].id, forKey: activeProgramDefaultsKey)
    }

    func programsInProfile(from catalog: [WorkoutProgram]) -> [WorkoutProgram] {
        let allowed = idsInLibrary(catalogIds: catalog.map(\.id))
        return catalog.filter { allowed.contains($0.id) }
    }
}

File: bp_workout_flutter/lib/providers/user_program_library.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

/// Same keys as iOS `UserProgramLibrary.swift`.
final userProgramLibraryProvider =
    NotifierProvider<UserProgramLibraryNotifier, int>(
  UserProgramLibraryNotifier.new,
);

class UserProgramLibraryNotifier extends Notifier<int> {
  static const libraryKey = 'profile.libraryProgramIds';
  static const activeProgramKey = 'workoutHub.activeProgramId';

  @override
  int build() => 0;

  void _bump() => state++;

  /// Default “all catalog programs” when key is unset.
  Future<Set<String>> idsInLibrary(List<String> catalogIds) async {
    final p = await SharedPreferences.getInstance();
    final catalog = catalogIds.toSet();
    if (!p.containsKey(libraryKey)) return catalog;
    final stored = p.getStringList(libraryKey) ?? [];
    return stored.toSet().intersection(catalog);
  }

  Future<bool> isInLibrary(String programId, List<String> catalogIds) async {
    final s = await idsInLibrary(catalogIds);
    return s.contains(programId);
  }

  Future<void> setProgramInLibrary(
    String programId,
    bool enabled,
    List<String> catalogIds,
  ) async {
    final p = await SharedPreferences.getInstance();
    var next = await idsInLibrary(catalogIds);
    if (enabled) {
      next.add(programId);
    } else {
      next.remove(programId);
    }
    await p.setStringList(libraryKey, next.toList());
    await _reconcileActiveProgram(p, catalogIds, next);
    _bump();
  }

  /// Sets the hub’s active program (must be in profile). Same key as iOS.
  Future<void> setActiveProgramId(String? programId) async {
    final p = await SharedPreferences.getInstance();
    if (programId == null || programId.isEmpty) {
      await p.remove(activeProgramKey);
    } else {
      await p.setString(activeProgramKey, programId);
    }
    _bump();
  }

  Future<void> _reconcileActiveProgram(
    SharedPreferences p,
    List<String> catalogIds,
    Set<String> inLib,
  ) async {
    final inProfilePrograms = catalogIds.where(inLib.contains).toList();
    if (inProfilePrograms.isEmpty) {
      await p.setString(activeProgramKey, '');
      return;
    }
    final saved = p.getString(activeProgramKey);
    if (saved != null &&
        saved.isNotEmpty &&
        inProfilePrograms.contains(saved)) {
      return;
    }
    await p.setString(activeProgramKey, inProfilePrograms.first);
  }
}
